presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java

```diff
@@ -0,0 +1,137 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class FixedNumericHistogram
+{
+    private static final byte FORMAT_TAG = 0;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FixedNumericHistogram.class).instanceSize();
+
+    private final int numBuckets;
+    private final double min;
+    private final double max;
+    private final double[] weights;
+
+    public FixedNumericHistogram(int numBuckets, double min, double max)
+    {
+        checkArgument(numBuckets >= 2, "numBuckets must be >= 2");
+        checkArgument(min < max, "max must be larger than min");
+
+        this.numBuckets = numBuckets;
+        this.min = min;
+        this.max = max;
+        this.weights = new double[numBuckets];
+    }
+
+    public FixedNumericHistogram(SliceInput input)
+    {
+        checkArgument(input.readByte() == FORMAT_TAG, "Unsupported format tag");
+
+        numBuckets = input.readInt();
+        checkArgument(numBuckets >= 2, "numBuckets must be >= 2");
+        min = input.readDouble();
+        max = input.readDouble();
+        checkArgument(min < max, "max must be greater than min");
+        this.weights = new double[numBuckets];
+        input.readBytes(Slices.wrappedDoubleArray(weights), numBuckets * SizeOf.SIZE_OF_DOUBLE);
+    }
+
+    public int getRequiredBytesForSerialization()
+    {
+        return SizeOf.SIZE_OF_BYTE + // format
+                SizeOf.SIZE_OF_INT + // num buckets
+                SizeOf.SIZE_OF_DOUBLE + // min
+                SizeOf.SIZE_OF_DOUBLE + // max
+                SizeOf.SIZE_OF_DOUBLE * numBuckets; // weights
+    }
+
+    public void serialize(SliceOutput out)
+    {
+        out.appendByte(FORMAT_TAG)
+            .appendInt(numBuckets)
+            .appendDouble(min)
+            .appendDouble(max)
+            .appendBytes(Slices.wrappedDoubleArray(weights, 0, numBuckets));
+    }
+
+    public long estimatedInMemorySize()
+    {
+        return INSTANCE_SIZE + SizeOf.sizeOf(weights);
+    }
+
+    public void set(double value, double weight)
+    {
+        checkArgument(value >= min, "value must not be smaller than min");
+        checkArgument(value <= max, "value must not be larger than max");
+        final int ind = Math.min(
```

Comment:
I would have a little helper function for here and add and call it something like `getIndex`